[SDMEXT-1040]: Support Secondary Properties In Attachments

PujaDeshmukh17 opened the pull request:
This PR is intended to implement the feature requirement of supporting custom properties in attachments. 

Jira Link: https://jira.tools.sap/browse/SDMEXT-1040
Wiki Link: https://wiki.one.int.sap/wiki/display/sapecm/Support+Secondary+Properties+In+Attachments

PujaDeshmukh17 committed:
Changes for CreateAttachmentHandler

PujaDeshmukh17 committed:
Code for updateAttachmentHandler

PujaDeshmukh17 committed:
Refactored helper methods to util class

PujaDeshmukh17 committed:
Refactor

PujaDeshmukh17 committed:
Adding new parameter to rename method

PujaDeshmukh17 committed:
Correcting tests in SDMCreateAttachmentHandlerTest

PujaDeshmukh17 committed:
Moving reuseable code to utils

PujaDeshmukh17 committed:
Adapting exsisting UTs are per code change

PujaDeshmukh17 committed:
Removing rm_hold check

PujaDeshmukh17 committed:
Reverting changes for createAttachmentHandler

PujaDeshmukh17 committed:
Working changes with mcm

PujaDeshmukh17 committed:
Exsisting UT fix

rishikunnath2747 committed:
changes for create flow

PujaDeshmukh17 committed:
UT

rishikunnath2747 committed:
Merge branch 'CustomPropertiesImplCreateFlow', remote-tracking branch 'origin' into CustomPropFinalBranch

rishikunnath2747 committed:
check for filename

rishikunnath2747 committed:
create flow name update

rishikunnath2747 committed:
fix

rishikunnath2747 committed:
incorrect fields check

rishikunnath2747 committed:
refactoring to change name

rishikunnath2747 committed:
unecessary types removal

rishikunnath2747 committed:
adding caching

PujaDeshmukh17 committed:
SDMUtils test

PujaDeshmukh17 committed:
SDMUpdateAttachmentHandler UTs

rishikunnath2747 committed:
fix for boolean

rishikunnath2747 committed:
removing properties cache

rishikunnath2747 committed:
removing comments

rishikunnath2747 committed:
fix date type

PujaDeshmukh17 committed:
SDMUpdateAttachmentHandler test

rishikunnath2747 committed:
Revert "SDMUpdateAttachmentHandler test"

This reverts commit 236b0d4e53ed7ced5da74d82bd5758350c0c34e2.

rishikunnath2747 committed:
Revert "fix date type"

This reverts commit 550722c9721462f863d24313490e309e9b7b27bb.

rishikunnath2747 committed:
annotation and check

rishikunnath2747 committed:
CreateAttachmentsHandlerUTs

PujaDeshmukh17 committed:
UpdateAttachmentHandler test

PujaDeshmukh17 committed:
Removing print statements

PujaDeshmukh17 committed:
Utils tests to increase coverage

rishikunnath2747 committed:
Fix

Improving test coverage and fixing annotation

rishikunnath2747 committed:
adding readme content

PujaDeshmukh17 committed:
Uts for sdmUpdateAttachmentHandler

rishikunnath2747 committed:
readme fix

rashmiangadi11 commented on sdm/src/main/java/com/sap/cds/sdm/caching/CacheConfig.java:
```diff
@@ -63,6 +65,17 @@ public static void initializeCache() {
+                    ResourcePoolsBuilder.heap(HEAP_SIZE))
+                .withExpiry(
+                    Expirations.timeToLiveExpiration(
+                        new Duration(ACCESS_TOKEN_EXPIRY, TimeUnit.MINUTES))));
```
Can u create a new variable to maintain this expiry

rishikunnath2747 replied:
Set it to not expire

rashmiangadi11 commented on sdm/src/main/java/com/sap/cds/sdm/handler/applicationservice/SDMCreateAttachmentsHandler.java:
```diff
@@ -82,6 +89,11 @@ private void processAttachment(
       List<String> fileNameWithRestrictedCharacters,
       List<String> duplicateFileNameList)
       throws IOException {
+    String id = (String) attachment.get("ID"); // Ensure appropriate cast to String
```
will this be always ID or can it be Id

rishikunnath2747 replied:
It will always be ID

rashmiangadi11 commented on sdm/src/main/java/com/sap/cds/sdm/handler/applicationservice/SDMUpdateAttachmentsHandler.java:
```diff
@@ -108,33 +106,54 @@ public void processAttachment(
         case 409:
           duplicateFileNameList.add(filenameInRequest);
-          attachment.replace("fileName", fileNameInSDM);
+          // attachment.replace("fileName", fileNameInSDM);
```
can we remove this commented line

PujaDeshmukh17 replied:
Done.

rashmiangadi11 commented on sdm/src/main/java/com/sap/cds/sdm/service/SDMService.java:
```diff
@@ -41,9 +43,23 @@ public void readDocument(
+  public List<String> getSecondaryTypes(
+      String repositoryId, String jwtToken, SDMCredentials sdmCredentials) throws IOException;
+
+  public List<String> getValidSecondaryProperties(
```
why is this added it can be just a normal method instead of adding in interface

rashmiangadi11 replied:
but implementation is same can we keep in SDMUtils

PujaDeshmukh17 replied:
Since these methods are part of the core functionality and represents a business operation we have kept it in the service class.

rishikunnath2747 committed:
error msg

rashmiangadi11 commented on sdm/src/main/java/com/sap/cds/sdm/service/SDMServiceImpl.java:
```diff
@@ -404,4 +458,78 @@ public int deleteDocument(String cmisaction, String objectId, String userEmail,
+        }
+        return result;
+      } catch (IOException e) {
+        throw new ServiceException("Could not update the attachment", e);
```
can u put the error message in SDMConstants

PujaDeshmukh17 replied:
Done.

rashmiangadi11 commented on sdm/src/main/java/com/sap/cds/sdm/service/SDMServiceImpl.java:
```diff
@@ -127,25 +133,73 @@ private void formResponse(
+      if (response.getStatusLine().getStatusCode() == 400
+          && responseContent.contains("is unknown!")) {
+        throw new ServiceException(
+            "Unknown secondary property, kindly verify all the secondary properties");
```
Put this message in SDMConstants

rishikunnath2747 replied:
Done

rashmiangadi11 commented on sdm/src/main/java/com/sap/cds/sdm/service/SDMServiceImpl.java:
```diff
@@ -404,4 +458,78 @@ public int deleteDocument(String cmisaction, String objectId, String userEmail,
+    if (secondaryTypes == null) {
+      String subdomain = TokenHandler.getSubdomainFromToken(jwtToken);
+      var httpClient =
+          TokenHandler.getHttpClient(binding, connectionPool, subdomain, "TOKEN_EXCHANGE");
```
is this user specific api or admin api

rishikunnath2747 replied:
user specific api

rishikunnath2747 committed:
fix ut + readme

PujaDeshmukh17 committed:
SDMServiceImp UT

rishikunnath2747 committed:
Revert "SDMServiceImp UT"

This reverts commit f5a9ba5f1ccfc3ba18f3878a4ac4129fac07aaa5.

rishikunnath2747 committed:
UT change

PujaDeshmukh17 committed:
SDMServiceImpl test

PujaDeshmukh17 committed:
Review comments

PujaDeshmukh17 committed:
Adding error as constant

rishikunnath2747 committed:
improving ut coverage

rishikunnath2747 committed:
ut fix

rishikunnath2747 committed:
ut changes

rishikunnath2747 committed:
sonar fixes

rishikunnath2747 committed:
removing test

rishikunnath2747 committed:
Update cache

Setting caching for secondary types as infinite as it only needs to be refreshed on application redeployment

rishikunnath2747 committed:
excluding class from sonar scan

maheshsooryambylu reviewed (approved):
Went through the code changes. Code looks fine to me.

rishikunnath2747 committed:
fix ut

rishikunnath2747 committed:
fixes

rishikunnath2747 committed:
uts

commenting uts which work locally but not in github job

rishikunnath2747 committed:
spotless fix

rishikunnath2747 committed:
update filename and ut coverage

rishikunnath2747 committed:
resolving conflict

rishikunnath2747 committed:
fix spotless

rishikunnath2747 committed:
Merge branch 'develop' into CustomPropFinalBranch

rishikunnath2747 committed:
sonar fixes

rishikunnath2747 committed:
sonar fixes

rishikunnath2747 committed:
commenting ut

will be fixed later

rishikunnath2747 committed:
sonar fix

PujaDeshmukh17 commented:
As discussed, the UT branch coverage has been reduced to 80% which would be covered in the later PR.
